### ISSUE TITLE
Spoke agent for MCS controller

### ISSUE THREAD

**Issue by elevran** [axon-submariner/lighthouse#1]
# Active Hub Support in Lighthouse Agent

Refactor Lighthouse agent workflows to rely on an active MCS controller at the hub cluster (broker) (see #2) instead of a logic-less hub.

Please refer to this [doc](https://github.com/axon-net/lighthouse/blob/devel/docs/lh-workflows.md) for details on current agent behavior.

## Required Steps

- [x] High-level design

- [x] `serviceExportSyncer`:
    - [x] shall upload local `serviceexports` from service namespace to the broker namespace (uni-directional)
    - [x] Rename service export to `<svc-name>-<svc-ns>-<cluster-id>` on broker.
    - [x] Also if export is created out-of-band (when agent controller is down)
    - [x] Upload export only if service exist
    - [x] Delete export on broker if service is deleted. Recreate export on broker if service is recreated.  
       - [x] deleted and recreated online 
       - [x] deleted out-of-band and created online
       - [x] deleted online and created out-of-band
    - [x] Add labels and annotations to hub's copy to preserve original namespace, service name and cluster-id.
    - [x] Add  annotations to preserve service network details: IPs, ports, session affinity, type.
    - [x] Shall download export status from hub (e.g. conflict) and update (merge) into service status at spoke
       - [x] Also when broker's copy set to conflict out-of-band - not required because agent will re-upload all exports and this will cause recalculation of conflict state on broker exports by the hub.
    - [x] Shall preserve original export creation timestamp as a label (for conflict resolution)
    - [x] Consider using a config map instead of service export to avoid many annotations (a CRD will be introduced later) - decided to stay with a service export
    - [x] Shall delete export on broker if local export is deleted
       - [x] online
       - [x] out-of-band

- [x] `serviceImportDownloader` shall download remote `serviceimports` (created by the hub controller) from the broker namespace to `submariner-operator` namespace (uni-directional)
    - [x] Labels and annotations shall be preserved.
    - [x] Handle deletion of imports in broker out-of-band: delete local imports as well
    - [x] Handle creation of imports on broker out-of-band
    - [x] Download also imports that were created in the hub for exports of the local cluster

- [x] `serviceSyncer`:
    - [x] shall update service export labels upon a change in service network details. Check if currently lighthouse is doing this. If it does, no need to implement at the first iteration. Opened separate issue #6 
    - [x] shall update service export status on service availability change (deletion)

- [x] `endpointSliceSyncer` - shall upload local endpoint slices to broker + download remote endpoint slices from broker to service NS. Syncs only EPS that are managed by submariner.
   - [x] Shall publish EndpointSlice only if export is not in conflict (and delete hub's copy if its exist while in conflict) - this is done inherently because hub deletes imports for exports in conflict. Service import controller will be notified for import deletion and delete local EP slice, which will get synced to the broker.

- [x] `serviceImportController` - shall create an Endpoint controller for each *local* service import. Endpoint controller creates an endpoint slice for each endpoints object.  
    - [x] Consider changing the trigger to create endpoint controller from a creation of service import to creation of service export - opened separate issue #7. 

- [x] Update unit tests

- [ ] Update e2e tests

**Comment by elevran** [comment]
- agree that `<svc>`-`<ns>`-`<cluster>` is descriptive, do mind that there could be a name limitation (63 chars?) so maybe picking a fixed UUID would be safer? We still have the information needed as labels for reference.
- need to update the `ServiceExport.Status` (e.g., conflict with other exports on the hub). Not sure this is currently handled
- should not pull/create `ServiceImport` if the local spoke cluster does not have the relevant namespace (currently this is a bug in MCS handling by Submariner)
- the hub creates an import with the "target" namespace and name as labels/annotations. When downloading, these should be removed before creating in the local cluster, specifically if not using the submariner-operator namespace but the real ns.
- why are we syncing to the `submariner-operator` namespace and not to the relevant namespace?
- does the spoke expect a combined/merged `ServiceImport` or multiple imports (one per exporting cluster) that it will combine? Same applies to `EndpointSlice`: who does the merging from multiple clusters - the hub or the spoke?
- in *Unchanged* section: `serviceSyncer shall update service export status...` - I think you mean deletion mostly, but we may need to update the labels on the `ServiceExport` in response to changes in the `Service`. I don't think you mean "shall update `ServiceExport.Status` field", but it could be assumed from the wording.
- unclear about the use of Endpoint controller? Do we need one per export/import or one per cluster? I'm unclear what is its role and responsibility on the exporting/importing spoke given the `endpointSliceSyncer`?

**Comment by davidohana** [comment]
> * agree that `<svc>`-`<ns>`-`<cluster>` is descriptive, do mind that there could be a name limitation (63 chars?) so maybe picking a fixed UUID would be safer? We still have the information needed as labels for reference.

That's the current naming convention in lighthouse for service imports. I think we should keep it as it at first iteration.

**Comment by davidohana** [comment]
> should not pull/create ServiceImport if the local spoke cluster does not have the relevant namespace (currently this is a bug in MCS handling by Submariner)

Agreed. However It doesn't really matter from functional perspective if service import exist, because relevant endpoint slice is not created in service NS until NS exists. 
Actually it might be that for contacting a remote service all you need is the endpoint slice and it will work even without an import. Will check.

**Comment by davidohana** [comment]
> the hub creates an import with the "target" namespace and name as labels/annotations. When downloading, these should be removed before creating in the local cluster, specifically if not using the submariner-operator namespace but the real ns.

Import  is currently created in the submariner-operator ns, so we should preserve those labels.

**Comment by davidohana** [comment]
> why are we syncing to the submariner-operator namespace and not to the relevant namespace?

Preserving current behavior. Perhaps so that users will not delete it? Not to clutter their user namepace? Can ask submariner team why they did that ..

**Comment by davidohana** [comment]
> does the spoke expect a combined/merged ServiceImport or multiple imports (one per exporting cluster) that it will combine? Same applies to EndpointSlice: who does the merging from multiple clusters - the hub or the spoke?

service import and endpoint slice per cluster. No persistent merge is done, probably merge is in memory.

**Comment by davidohana** [comment]
> unclear about the use of Endpoint controller? Do we need one per export/import or one per cluster? I'm unclear what is its role and responsibility on the exporting/importing spoke given the endpointSliceSyncer?

Currently, there's an endpoint controller per **local** service import.  It creates the endpoint slice that is synced by endpointSliceSyncer

**Comment by davidohana** [comment]
> need to update the ServiceExport.Status (e.g., conflict with other exports on the hub). Not sure this is currently handled

I am also not sure how network conflicts are currently handled. Will check.
In the new design - which of the exports should be in conflict? The older one? all? new one?
Also I think that conflict shall be declared by the hub not the spoke, so its not in this task scope.

**Comment by davidohana** [comment]
>  in *Unchanged* section: `serviceSyncer shall update service export status...` - I think you mean deletion mostly, but we may need to update the labels on the `ServiceExport` in response to changes in the `Service`. I don't think you mean "shall update `ServiceExport.Status` field", but it could be assumed from the wording.

I meant updating the status field. This is the current behavior and it ends up in deletion of the import (not clear how, I need to understand better how this works.. )

**Comment by elevran** [comment]
> Agreed. However It doesn't really matter from functional perspective if service import exist, because relevant endpoint slice is not created in service NS until NS exists.
The service name (with the `clusterset.local`) becomes resolvable to an IP once the Import is created. It shouldn't appear at all.

> In the new design - which of the exports should be in conflict? The older one? all? new one?
I would vote for the oldest. `Conflict` status should also report what the conflict is (which can be determined from comparing the new Export with a previous one.

**Comment by elevran** [comment]
I'm assuming that for the initial change we would want to minimize the changes and keep it backwards compatible. Correct?
Future iterations would consider changes (e.g., Imports must be merged in MCS, AFAIU; use of operator namespace, ...).

**Comment by davidohana** [comment]
> > * agree that `<svc>`-`<ns>`-`<cluster>` is descriptive, do mind that there could be a name limitation (63 chars?) so maybe picking a fixed UUID would be safer? We still have the information needed as labels for reference.
> 
> That's the current naming convention in lighthouse for service imports. I think we should keep it as it at first iteration.

From what I read, resource name limitation is 253 chars and not 63: https://unofficial-kubernetes.readthedocs.io/en/latest/concepts/overview/working-with-objects/names/#:~:text=Names%20are%20used%20to%20refer,resources%20have%20more%20specific%20restrictions.
